crates/cext/src/transpiler/target.rs

```diff
@@ -0,0 +1,777 @@
+            map: Default::default(),
+        }
+    }
+
+    pub fn new_fixed(operation: StandardGate, params: SmallVec<[Param; 3]>) -> Self {
+        Self {
+            operation,
+            params: Some(params),
+            map: Default::default(),
+        }
+    }
+}
+
+/// @ingroup QkTargetEntry
+/// Creates an entry to the ``QkTarget`` based on a ``QkGate`` instance with
+/// no parameters.
+///
+/// @param operation The Standard Gate representing this entry in the target.
+///
+/// @return The ``QkTargetEntry`` structure.
+///
+/// # Example
+///
+///     QkTargetEntry *entry = qk_target_entry_new(QkGate_H);
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_target_entry_new(operation: StandardGate) -> *mut TargetEntry {
+    // Fast fail if the instruction is expecting parameters.
+    if operation.num_params() != 0 {
+        panic!("Tried to create an non-parametric entry with a parametric gate.")
+    }
+    Box::into_raw(Box::new(TargetEntry::new(operation)))
+}
+
+/// @ingroup QkTargetEntry
+/// Creates an object that will serve as a mapping between an instruction's
+/// qargs and instruction properties.
+///
+/// @param operation The Standard Gate representing this entry in the target.
+/// @param params The list of fixed parameters for this gate.
+///
+/// @return The ``QkTargetEntry`` structure.
+///
+/// # Example
+///
+///     double crx_params[1] = {3.14};
+///     QkTargetEntry *entry = qk_target_entry_new(QkGate_CRX, crx_params);
+///
+/// # Safety
+///
+/// The ``params`` type is expected to be a pointer to an array of ``double`` where the length
+/// matches the the expectations of the standard gate. If the array is insufficently long the
+/// behavior of this function is undefined as this will read outside the bounds of the array.
+/// It can be a null pointer if there are no params for a given gate. You can check
+/// `qk_gate_num_params` to determine how many qubits are required for a given gate.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_entry_new_fixed(
+    operation: StandardGate,
+    params: *mut f64,
+) -> *mut TargetEntry {
+    unsafe {
+        Box::into_raw(Box::new(TargetEntry::new_fixed(
+            operation,
+            parse_params(operation, params),
+        )))
+    }
+}
+
+/// @ingroup QkTargetEntry
+/// Retrieves the length of the current target entry's property map..
+///
+/// @param entry The pointer to the mapping object.
+///
+/// @return The length of the TargetEntry.
+///
+/// # Example
+///
+///     // Create an entry for an H gate
+///     QkTargetEntry *entry = qk_target_entry_new(QkGate_H);
+///     size_t props_size = qk_target_entry_len(entry);
+///
+/// # Safety
+///
+/// The behavior is undefined if ``entry`` is not a valid,
+/// non-null pointer to a ``QkTargetEntry`` object.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_entry_len(entry: *const TargetEntry) -> usize {
```

Comment:
Also here, could we call it something that reflects the property better, like `num_instructions` or something? 🙂

Reply:
For an entry, I think the current name is valid since it is the number of both qargs and properties (we could get away with calling it `qk_target_entry_num_props` or something among those lines). This name, however, could work for the `qk_target_len` method.